### PR TITLE
Support external custom operator schemas on Ubuntu

### PR DESCRIPTION
**Pull request by satyajandhyala**
**Description**: Describe your changes.
_External custom ops_: Unlike custom operations defined in the onnxruntime code base, the external custom op schemas are compiled into a shared library independently and loaded at runtime. The user should be able to create and use custom ops using prebuilt onnxruntime. The user doesn't need to build onnxruntime to use the external custom ops.

This feature allows the user to directly specify custom op schemas in C/C++ and compile the code into a shared library that may be loaded into python script. See onnxruntime/orttraining/orttraing/test/external_custom_ops for example.

The custom ops need to register with registry in onnxruntime. In order to enable resolving onnx::* and protobuf::* symbols at load-time, onnxruntime needs to expose these symbols. The user code containing custom op schemas may be compiled using the ONNX header files bundled in the onnxruntime wheel. In addition to generated ONNX header files, the wheel contains Protobuf header files.

This PR doesn't add ORT API header files to onnxruntime wheel.

The current changes only support the use-case on Ubuntu platform. This feature requires building onnxruntime with --enable_external_custom_op_schemas and the code changes in this PR are guarded by this flag.

For more details please refer to https://microsoft-my.sharepoint.com/:w:/p/sajandhy/EbwOyIuMyUBEntJYYbjDLtoBDl53eNXlqKV4vfGxNBXb9A
**Motivation and Context**
This use-case is required for 1P users
- Why is this change required? What problem does it solve?
This change enables user to define operator schemas, compile and load into onnxruntime.

- If it fixes an open issue, please link to the issue here.
https://msdata.visualstudio.com/Vienna/_workitems/edit/1271875

### CODE REVIEW

**Review by SherlockNoMad — DISMISSED** [review]

File: setup.py [hunk 1]
```diff
@@ -254,6 +255,11 @@ def run(self):
 with open(README) as f:
     long_description = f.read()
 
+# Include files
+# Gather all files under onnxruntime/external directory.
+extra.extend(list(str(Path(*Path(x).parts[1:])) for x in list(iglob(
```
[the remark this answers is not in the document]
Comment (satyajandhyala) [reply setup.py]: Added an extra condition to make it clear that extra gets extended only if onnxruntime/external exists under the build binary directory.

File: cmake/onnxruntime_python.cmake [hunk 2]
```diff
@@ -157,7 +161,20 @@ else()
   set_property(TARGET onnxruntime_pybind11_state APPEND_STRING PROPERTY LINK_FLAGS " -Xlinker -rpath=\\$ORIGIN")
 endif()
 
-target_link_libraries(onnxruntime_pybind11_state PRIVATE ${onnxruntime_EXTERNAL_LIBRARIES})
+if (onnxruntime_ENABLE_EXTERNAL_CUSTOM_OP_SCHEMAS)
+  set(onnxruntime_CUSTOM_EXTERNAL_LIBRARIES "${onnxruntime_EXTERNAL_LIBRARIES}")
+  list(FIND  onnxruntime_CUSTOM_EXTERNAL_LIBRARIES onnx ONNX_INDEX)
+  list(FIND onnxruntime_CUSTOM_EXTERNAL_LIBRARIES ${PROTOBUF_LIB} PROTOBUF_INDEX)
+  MATH(EXPR PROTOBUF_INDEX_NEXT "${PROTOBUF_INDEX} + 1")
```
Comment (pranavsharma) [inline cmake/onnxruntime_python.cmake:168]: If PROTOBUF_INDEX = -1, are the next set of lines still valid?

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -544,6 +544,10 @@ def parse_arguments():
     parser.add_argument('--eager_customop_header', default=None,
                         help='Header containing custom op definitions for eager mode.')
 
+    parser.add_argument(
```
Comment (pranavsharma) [inline tools/ci_build/build.py:547]: Should this be exposed in build.py given that this feature is going to be used only internally and also given the limitations the feature has (might not work on Windows, etc)?

File: cmake/onnxruntime_python.cmake (hunk 2 above)
Comment (satyajandhyala) [reply cmake/onnxruntime_python.cmake:168]: Yes. it is still valid. But we don't use it.

File: tools/ci_build/build.py (hunk 3 above)
Comment (satyajandhyala) [reply tools/ci_build/build.py:547]: I don't think it should cause any problem if we document that it only works on Ubuntu.

File: cmake/onnxruntime_python.cmake (hunk 2 above)
Comment (satyajandhyala) [reply cmake/onnxruntime_python.cmake:168]: Fixed the logical error.

File: tools/ci_build/build.py (hunk 3 above)
Comment (satyajandhyala) [reply tools/ci_build/build.py:547]: Added a fix in build.py to fail if the platform is not Ubuntu if --enable_external_custom_op_schemas option is specified.

**Review by SherlockNoMad — APPROVED** [review]

**Review by ashari4 — COMMENTED** [review]
I am removing myself as a reviewer to avoid blocking these changes from getting in sooner.

**Review by SherlockNoMad — APPROVED** [review]

**Pull request merged**